src/mds/Server.cc

```diff
@@ -10927,6 +10927,7 @@ void Server::handle_client_mksnap(MDRequestRef& mdr)
   pi.inode->ctime = info.stamp;
   if (info.stamp > pi.inode->rstat.rctime)
     pi.inode->rstat.rctime = info.stamp;
+  pi.inode->change_attr++;
```

Comment:
Are we incrementing the change_attr of the .snap directory's parent  (e.g., In case of  `mkdir dir00/.snap/snap00`, "_dir00_" is .snap directory's parent) here? As per https://github.com/ceph/ceph/pull/48382#issuecomment-1274125098 I thought we didn't want to do that.  

Besides storing `last_modified` in the snaprealm, maybe we also want to store a `change_attribute`?

Reply:
Is there any good reason why we have to do that ? The `sr_t` is just a metadata info in the parent directory and IMO the `last_modified` should have helped track it.

Reply:
Doesn't NFS rely on both mtime and change_attr being modified for its cache invalidation? @ajarr

Reply:
> Besides storing `last_modified` in the snaprealm, maybe we also want to store a `change_attribute`?

What `change_attribute` are we talking about? Could you elaborate?

Reply:
> > Besides storing `last_modified` in the snaprealm, maybe we also want to store a `change_attribute`?
> 
> What `change_attribute` are we talking about? Could you elaborate?

I was talking about `change_attribute` for  the .snap directory, which AFAIU is modeled as a snap realm in the MDS.  

As Xiubo suggests in https://github.com/ceph/ceph/pull/48382/files#r1008957221, maybe we just rely on snap realm's `last_modified` (basically mtime that changes with creation, deletion, or removal of snaps within the realm)?  We can do that, and convert last_modified (utime_t) to change_attr (uint64_t) on the client side. Anyways, I don't think we want to increment the change_attr of .snap directory's parent considering the discussion here, https://github.com/ceph/ceph/pull/48382#issuecomment-1274118916 .  We don't want to unnecessarily increment a directory's change_attr that will invalidate NFS kclient's readdir cache unnecessarily and lower performance.

Reply:
> Is there any good reason why we have to do that ? The `sr_t` is just a metadata info in the parent directory and IMO the `last_modified` should have helped track it.

Yes, we can use the snap realm's `last_modified` to generate change_attr on the client side for .snap directory. Maybe we should just do this.  

This was my thinking: Wouldn't the same reason for introducing change_attr in a regular MDS inode hold here too? We could've relied on an inode's mtime/ctime changes to generate NFSv4 change_attr on the client side for the inode.

Reply:
> > Is there any good reason why we have to do that ? The `sr_t` is just a metadata info in the parent directory and IMO the `last_modified` should have helped track it.
> 
> Yes, we can use the snap realm's `last_modified` to generate change_attr on the client side for .snap directory. Maybe we should just do this.

Sounds good to me.

> 
> This was my thinking: Wouldn't the same reason for introducing change_attr in a regular MDS inode hold here too? We could've relied on an inode's mtime/ctime changes to generate NFSv4 change_attr on the client side for the inode.

Maybe, unless `change_attr` is supposed to be monotonically increasing. (users can set inode's {a,m}time - if `change_attr` is derived from mtime, then it _go back_ (in time?).

Reply:
@ajarr One issue that can crop up when using mtime as change_attr for snapdir is that in the MDS `mdr->get_op_stamp()` is used to set `last_modified` for a snap realm - https://github.com/ceph/ceph/pull/48382#discussion_r1006439332

However, we have this issue in many places, so its nothing new.

Reply:
@ajarr @lxbsz I take that back -- we might loose precision when converting utime_t to uint64_t (version). I guess we are better off introducing `change_attr` in sr_t along with `last_modified`.

Reply:
> @ajarr @lxbsz I take that back -- we might loose precision when converting utime_t to uint64_t (version). I guess we are better off introducing `change_attr` in sr_t along with `last_modified`.

Sounds reasonable.